lib/mempool/indexer.js

```diff
@@ -0,0 +1,481 @@
+      this.coinIndex.insert(tx, i);
+  }
+
+  /**
+   * Transaction was removed from mempool.
+   * This might happen for several reasons:
+   *  - Mempool size limit got rid of it.
+   *  - Transaction was included in block.
+   *  - Double spend in a block
+   *  - After reorg tx is no longer final
+   * We concentrate on recovering inputs as coins.
+   * @param {MempoolEntry} entry
+   */
+
+  removeEntry(entry) {
+    const tx = entry.tx;
+    const hash = tx.hash();
+
+    this.txIndex.remove(hash);
+
+    for (const {prevout} of tx.inputs) {
+      const {hash, index} = prevout;
+      const prev = this.mempool.getTX(hash);
+
+      if (!prev)
+        continue;
+
+      this.coinIndex.insert(prev, index);
+    }
+
+    for (let i = 0; i < tx.outputs.length; i++)
+      this.coinIndex.remove(hash, i);
+  }
+
+  /**
+   * Block disconnected and we recover coins if they are available.
+   * This event comes after `tx` and `add entry`, we might want to
+   * check if outputs indexed by those are already spent in the mempool.
+   * @param {TX} tx
+   * @param {Block} block
+   */
+
+  unconfirmed(tx, block) {
+    const hash = tx.hash();
+
+    for (let i = 0; i < tx.outputs.length; i++) {
+      if (this.mempool.isSpent(hash, i))
+        this.coinIndex.remove(hash, i);
+    }
+  }
+
+  /**
+   * Reset indexes
+   * @private
+   */
+
+  reset() {
+    this.txIndex.reset();
+    this.coinIndex.reset();
+  }
+
+  /**
+   * Find all transactions pertaining to a certain address.
+   * Note: this does not accept multiple addresses.
+   * @param {Address} addrs
+   * @returns {TX[]}
+   */
+
+  getTXByAddress(addrs) {
+    if (!Array.isArray(addrs))
+      addrs = [addrs];
+
+    const out = [];
+
+    for (const addr of addrs) {
+      const hash = Address.getHash(addr, this.network);
+      const txs = this.txIndex.get(hash);
+
+      for (const tx of txs)
+        out.push(tx);
+    }
+
+    return out;
+  }
+
+  /**
+   * Find all transactions pertaining to a certain address.
+   * @param {Address} addrs
+   * @param {TXMeta[]]}
+   */
+
+  getMetaByAddress(addrs) {
+    if (!Array.isArray(addrs))
+      addrs = [addrs];
+
+    const out = [];
+
+    for (const addr of addrs) {
+      const hash = Address.getHash(addr);
+      const txs = this.txIndex.getMeta(hash);
+
+      for (const tx of txs)
+        out.push(tx);
+    }
+
+    return out;
+  }
+
+  /**
+   * Find all coins pertaining to a certain address.
+   * @param {Address} addr
+   * @return {Coin[]}
+   */
+
+  getCoinsByAddress(addrs) {
+    if (!Array.isArray(addrs))
+      addrs = [addrs];
+
+    const out = [];
+
+    for (const addr of addrs) {
+      const hash = Address.getHash(addr);
+      const coins = this.coinIndex.get(hash);
+
+      for (const coin of coins)
+        out.push(coin);
+    }
+
+    return out;
+  }
+}
+
+/**
+ * Mempool Indexer Options
+ * @alias module:mempool.MempoolIndexerOptions
+ */
+
+class MempoolIndexerOptions {
```

Comment:
These options don't seem necessary since there is really only 1 option that it sets